File: lintcode/intersection-of-two-linked-lists.py
#!/usr/bin/env python
# encoding: utf-8

# Question: http://www.lintcode.com/en/problem/intersection-of-two-linked-lists/

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param headA: the first list
    # @param headB: the second list
    # @return: a ListNode
    def getIntersectionNode(self, headA, headB):
        # Write your code here
        if not headA or not headB:
            return None

        cursor = headA
        while cursor.next:
            cursor = cursor.next
        cursor.next = headB

        cursor = headA
        fast = headA
        while fast and fast.next:
            cursor = cursor.next
            fast = fast.next.next
            if cursor == fast:
                break
        if cursor == fast:
            another_cursor = headA
            while cursor != another_cursor:
                cursor = cursor.next
                another_cursor = another_cursor.next
            return cursor
        return None

File: leetcode/intersection-of-two-linked-lists.cc
// Question: https://leetcode.com/problems/intersection-of-two-linked-lists/

#include "headers.h"

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        if (!headA || !headB) {
            return NULL;
        }

        ListNode* c = headA;
        while (c->next) {
            c = c->next;
        }
        c->next = headB;

        ListNode* fast = headA;
        ListNode* slow = headA;
        ListNode* entry = NULL;
        while (fast && fast->next) {
            slow = slow->next;
            fast = fast->next->next;
            if (slow == fast) {
                entry = headA;
                while (slow != entry) {
                    slow = slow->next;
                    entry = entry->next;
                }
                c->next = NULL;
                return entry;
            }
        }
        c->next = NULL;
        return NULL;
    }
};
